File: Python/2468_안전영역.py
if __name__ == "__main__":
	N = int(input())
	
	area = []
	for i in range(N):
		area.append(list(map(int,input().split())))
		
	safe_area = 1 #아무 것도 안잠겼을 때는 안전영역의 수 1
	
	dir = [[-1,0],[1,0],[0,-1],[0,1]] # 상하좌우
	
	H = list(set(sum(area,[]))) # 지역 높이의 set(1이상 100이하를 모두 해볼 수는 없으니까)
	for h in H: # 높이만큼 돌면서 진행
		visit = [[False]*N for i in range(N)]
		tmp_area = 0
		for i in range(N): 
			for j in range(N): # 배열 돌면서 시작점 설정
				if area[i][j] > h and visit[i][j] == False:
					tmp_area += 1 # 구역 1 증가
					
					queue = []
					queue.append([i,j])
					visit[i][j] = True
					while queue: # 시작 지점과 연결된 지점들은 bfs로 방문하기
						n = queue.pop(0)
						
						for d in dir:
							x = n[0]+d[0]
							y = n[1]+d[1]
							if x >= 0 and x < N and y >= 0 and y < N and visit[x][y] == False and area[x][y] > h:
								queue.append([x,y])
								visit[x][y] = True
				
				elif area[i][j] <= h and visit[i][j] == False: #h보다 작거나 같으면 침수된 지역이기 때문에 방문했다고 생각하기
					visit[i][j] = True
		
		if tmp_area > safe_area: # 최댓값 구하기 위함
			safe_area = tmp_area
			
			
	print(safe_area)

File: C++/2468_안전영역.cpp
#include <iostream>
#include <set>
#include <queue>
#include <cstring>
using namespace std;

struct Pos{
	int x, y;
};

int main() {
   int N;
   cin >> N;
   
   int **area = new int*[N];
   for(int i = 0; i < N; i++){
      area[i] = new int[N];
   }
   
   set<int> H; // 높이에 관한 set
   for(int i = 0; i < N; i++){
      for(int j = 0; j < N; j++){
         int n;
         cin >> n;
         H.insert(n);
         area[i][j] = n;
      }
   }
   
   int num = 1; // 구역의 수 초기화(침수된 곳 없으면 1임)
   
   set<int>::iterator iter;
   for(iter = H.begin(); iter != H.end(); iter++){
   	bool **visit = new bool*[N]; //방문 여부 표시할 배열
   	for(int i = 0; i < N; i++){
      visit[i] = new bool[N];
      memset(visit[i],false,sizeof(bool)*N);
   }
   
   int dir[4][2] = {{-1,0},{1,0},{0,-1},{0,1}}; //상하좌우
   int tmp_num = 0; // 현재 높이까지 잠겼을 때의 구역 수
   	for(int i = 0; i < N; i++){
   		for(int j = 0; j < N; j++){
   			if(area[i][j] > *iter && visit[i][j] == false){ // 높이가 현재 침수 높이보다 높아야하며 방문한 적 없어야 함
   				tmp_num++;
   				visit[i][j] = true;
   				
   				queue<Pos> q;
   				q.push({i,j});
   				
   				while(!q.empty()){ // 현재 i,j를 시작으로 bfs(bfs로 연결되어 있으면 구역 하나)
   					Pos p = q.front();
   					q.pop();
   					
   					for(int d = 0; d < 4; d++){
   						int x = p.x + dir[d][0];
   						int y = p.y + dir[d][1];
   						if(x >= 0 && x < N && y >= 0 && y < N && visit[x][y] == false && area[x][y] > *iter){
   							visit[x][y] = true;
   							q.push({x,y});
   						}
   					}
   				}
   			}
   			else if(area[i][j] <= *iter && visit[i][j] == false){ // 침수지역이므로 false여도 방문한 셈 치기
   				visit[i][j] = true;
   			}
   		}
   	}
   	
   	if(num < tmp_num){ // 최댓값 구하는 과
   		num = tmp_num;
   	}
   }
   
   cout << num;
   
   return 0;
}